book/scheduling-and-threading.md

```diff
@@ -928,181 +918,112 @@ class Browser:
-annoying that you will quickly not want to use that browser. But don't take my
-word for it---let's implement an artificial slowdown and you can see for
-yourself.
-
-Add the slowdown to our counter page as follows, with a 200ms synchronous delay
-running JavaScript:
-
-``` {.javascript file=eventloop}
-var count = 0;
-var start_time = Date.now();
-var cur_frame_time = start_time;
-
-artificial_delay_ms = 200;
-
-function callback() {
-    var since_last_frame = Date.now() - cur_frame_time;
-    while (since_last_frame < artificial_delay_ms) {
-        var since_last_frame = Date.now() - cur_frame_time;
-    }
-    # ...
-}
-```
-
-To make the above code work, you'll need this small addition to the runtime 
-code to implement a subset of the [Date API][date-api]:
+[optimize-surfaces]: visual-effects.md#optimizing-surface-use
 
-[date-api]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date
+[skia-gpu]: https://skia.org/docs/user/api/skcanvas_creation/#gpu
 
-``` {.javascript}
-function Date() {}
-Date.now = function() {
-    return call_python("now");
-}
-```
-and Python bindings:
-``` {.python}
-class JSContext:
-    def __init__(self, tab):
-        # ...
-        self.interp.export_function("now",
-            self.now)
-
-    def now(self):
-        return int(time.time() * 1000)
-```
-
-Now load the page and hold down the down-arrow button. Observe how inconsistent
-and janky the scrolling is. Compare with no artificial delay, which is pleasant
-to scroll.
-
-Browsers can and do optimize their JavaScript engines to be as fast as possible,
-but this has its limits. So the only way to keep the browser
-guaranteed-responsive is to run key interactions in parallel with JavaScript.
-Luckily, it's pretty easy to use the raster and draw thread we're planning
-to *also* scroll and interact with browser chrome.
+::: {.further}
+Even with a second thread, the browser thread can wait up to 66ms
+before *starting* to handle a click event! For this reason, modern
+browsers use [*yet more*][renderingng-architecture] threads or
+processes. For example, raster-and-draw might run on its own thread,
+so that it can't block event handling. This makes the browser thread
+extremely responsive to input, at the cost of even more complexity.
+:::
 
-The browser thread
-==================
+[renderingng-architecture]: https://developer.chrome.com/blog/renderingng-architecture/#process-and-thread-structure
 
-This new thread will be called the *browser thread*.[^also-compositor] The
-browser thread will be for:
+Two threads
+===========
 
-[^also-compositor]: Th browser thread is similar to what modern browsers often
-call the [*compositor thread*][cc].
+Running rendering in parallel with raster and draw would allow us to
+produce a new frame every 66ms, instead of every 88ms. But more
+importantly, since there's no point to running render more often than
+raster-and-draw, the render thread would have 46ms left over for run
```

Comment:
Hmm, I think this is maybe too confusing. I meant that for each animation frame it would have 20ms of rendering time and 46ms of JS time. I'll rewrite.